Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Calibration/EditorExtrinsicsCalibration.cs

```diff
@@ -0,0 +1,305 @@
+
+        [Header("VR Headset Parameters")]
+        /// <summary>
+        /// Used to obtain headset information for a vr headset.
+        /// </summary>
+        [Tooltip("Used to obtain headset information for a vr headset.")]
+        [SerializeField]
+        private HeadsetCalibration headsetCalibration = null;
+
+        [Header("Universal Parameters")]
+        public RawImage feedImage;
+        public RawImage lastArUcoImage;
+        public DebugVisualHelper markerVisualHelper;
+        public DebugVisualHelper cameraVisualHelper;
+
+        private CalibrationAPI calibration = null;
+        private CalculatedCameraIntrinsics dslrIntrinsics;
+        private Queue<HeadsetCalibrationData> dataQueue = new Queue<HeadsetCalibrationData>();
+        private int nextArUcoImageId = 0;
+        private List<CalculatedCameraExtrinsics> cameraExtrinsics;
+        private CalculatedCameraExtrinsics globalExtrinsics;
+        private List<GameObject> parentVisuals = new List<GameObject>();
+        private int expectedMarkers = 18;
+
+#if UNITY_EDITOR
+        private void OnValidate()
+        {
+            FieldHelper.ValidateType<INetworkingService>(NetworkingService);
+            FieldHelper.ValidateType<IMatchMakingService>(MatchMakingService);
+        }
+
+        private void Start()
+        {
+            CalibrationDataHelper.Initialize(out var nextChessboardImageId, out nextArUcoImageId);
+            dslrIntrinsics = CalibrationDataHelper.LoadCameraIntrinsics(cameraIntrinsicsPath);
+            if (dslrIntrinsics == null)
+            {
+                throw new Exception("Failed to load the camera intrinsics file.");
+            }
+            else
+            {
+                Debug.Log($"Successfully loaded the provided camera intrinsics file: {dslrIntrinsics}");
+            }
+
+            calibration = new CalibrationAPI();
+
+            networkingService = NetworkingService as INetworkingService;
+            if (networkingService != null)
+            {
+                networkingService.DataReceived += OnDataReceived;
+            }
+
+            matchMakingService = MatchMakingService as IMatchMakingService;
+            if (matchMakingService != null)
+            {
+                matchMakingService.Connect();
+            }
+
+            if (headsetCalibration != null)
+            {
+                headsetCalibration.Updated += OnHeadsetCalibrationUpdated;
+            }
+
+            for(int i = 0; i < nextArUcoImageId; i++)
+            {
+                var dslrTexture = CalibrationDataHelper.LoadDSLRArUcoImage(i);
+                var headsetData = CalibrationDataHelper.LoadHeadsetData(i);
+
+                if (dslrTexture == null ||
+                    headsetData == null ||
+                    !ProcessArUcoData(headsetData, dslrTexture))
+                {
+                    Debug.LogWarning($"Failed to process dataset: {i}");
+                }
```

Comment:
How often do you expect these things to actually fail? It looks like if it does fail, we could end up with unused IDs in the middle, where the next time we do some operations we end up re-using old IDs because we're basing the next ID on the number of files in a directory, rather than the actual filenames themselves.

For example, imagine that we have 10 things, we only load the 10th (so let's say now there is file_9.png in the directory).

The next time this runs this will say that the nextID is 1, and will try to use ID space 1-11, and if it succeeds this time it will blow away the old.

Maybe this is fine, but anytime I see ID reuse that is scary.

If in doubt just use the millis timestamp since epoch or something, which is unique and easily understandable.

Reply:
i have it at seconds right now. we can change it to something more robust if needed, but we should now avoid using random indices.